File: List/list.cpp
#include <iostream>

template<typename T>
class List {
    private:
        // Num of items in list
        int num;

        // Capacity of the list
        int capacity;

        // Array that holds the 
        // list items 
        T* items;
    
    public:
        List(void);

        // Add a new value to the list
        void add(T);

        // Get the value at index n
        T get(int);

        // Remove the item at
        // index n 
        void remove(int);

        // Remove the given
        // item
        void remove(T*);

        // Return the size of the list
        int size(void);

        // Return true if list is empty
        bool isEmpty(void);

        // Whether or not the list
        // contains the given item
        bool contains(T*);

        // Returns the index of the
        // given element
        int indexOf(T);

        // Returns the index of
        // the given pointer
        // to an element
        int indexOf(T*);

        // Set the item at given idx
        // to the given item
        void set(int, T*);

        // Return first item
        T first(void);

        // Return last item
        T last(void);

        // Display contained items
        void display(void);

        // Return real capacity of list
        int cap(void);
};

int main(void) {
    List<int> l;

    std::cout << "Is empty: " << l.isEmpty() << "\n";

    for(int i = 0; i < 10; i++)
        l.add(i * i * i);

    l.display();

    std::cout << "Size: " << l.size() << "\n";
    
    l.remove(1);
    l.display();
    std::cout << "Size: " << l.size() << "\n";
    int x = 8;
    int* p = &x;
    std::cout << "Contains 8: " << l.contains(p) << "\n";
    std::cout << "First: " << l.first() << "\n";
    std::cout << "Last: " << l.last() << "\n";
    std::cout << "Real capacity: " << l.cap() << "\n";

    return 0;
};

template<typename T>
List<T>::List() {
    capacity = 0;
    num = 0;
    items = new T[++capacity];
};

template<typename T>
void List<T>::add(T v) {
    if(++num >= capacity) capacity *= 2;
    if(capacity > (sizeof(items) / sizeof(T))) {
        T* tmp = new T[capacity];
        for(int i = 0; i < num - 1; i++)
            tmp[i] = items[i];
        items = tmp;
    }
    items[num - 1] = v;
};

template<typename T>
void List<T>::display(void) {
    for(int i = 0; i < num; i++)
        std::cout << items[i] << " ";
    std::cout << "\n";
};

template<typename T>
int List<T>::size(void) {
    return num;
};

template<typename T>
T List<T>::get(int idx) {
    return idx <= num ? items[idx] : NULL;
};

template<typename T>
int List<T>::indexOf(T* t) {
    for(int i = 0; i < num; i++)
        if(*t == items[i]) return i;

    return -1;
};

template<typename T>
int List<T>::indexOf(T t) {
    return indexOf(&t);
};

template<typename T>
void List<T>::remove(int idx) {
    if(idx <= num) {
        for(int i = idx; i < num - 1; i++)
            items[i] = items[i + 1];
        items[--num] = items[capacity - 1];
    }
};

template<typename T>
void List<T>::remove(T* t) {
    int idx = indexOf(t);
    if(idx != -1) remove(idx);
};

template<typename T>
bool List<T>::isEmpty(void) {
    return num == 0;
};

template<typename T>
bool List<T>::contains(T* t) {
    return indexOf(t) != -1;
};

template<typename T>
T List<T>::first(void) {
    return isEmpty() ? NULL : items[0];
};

template<typename T>
T List<T>::last(void) {
    return isEmpty() ? NULL : items[num - 1];
};

template<typename T>
void List<T>::set(int idx, T* t) {
    if(idx <= num) items[idx] = *t;
};

template<typename T>
int List<T>::cap(void){
    return capacity;
}

File: List/list.rs
use std::{vec::Vec, fmt::Debug, cmp::PartialEq};

pub struct List<T> {
    /// Num of items in list
    size: usize,

    /// Capacity of the list
    capacity: usize,

    /// List of items
    items: Vec::<T>,
}

impl<T: Copy> List<T> where T: Debug + PartialEq {
    pub fn new() -> List<T> {
        List::<T> {
            size: 0,
            capacity: 1,
            items: Vec::<T>::with_capacity(1),
        }
    }

    pub fn add(&mut self, v: T) {
        // Check that new size fits capacity
        if self.size + 1 >= self.capacity {
            self.capacity *= 2;
        }

        // If capacity has changed, resize
        if self.capacity > self.items.len() {
            let mut tmp: Vec<T> = Vec::<T>::new();
            for i in 0..self.size {
                tmp.insert(i, self.items[i]);
            }

            self.items = tmp.clone();
        }

        // Assign new element and increase size
        self.items.insert(self.size, v);
        self.size += 1;
    }

    pub fn display(&self) {
        println!("{:?}", self.items);
    }

    pub fn size(&self) -> usize {
        self.size
    }

    pub fn get(&self, idx: usize) -> Option<T> {
        if idx <= self.size { 
            Some(self.items[idx]) 
        } else {
            None
        }
    }

    pub fn index_of_ref(&self, v: &T) -> i32 {
        let mut idx: i32 = -1;
        for i in 0..self.size {
            if *v == self.items[i] {
                idx = i as i32;
            }
        }

        idx
    }

    pub fn index_of_val(&self, v: T) -> i32 {
        self.index_of_ref(&v)
    }

    pub fn remove_by_idx(&mut self, idx: usize) {
        if idx <= self.size {
            for i in 0..self.size - 1 {
                self.items[i] = self.items[i + 1];
            }

            self.size -= 1;
            self.items[self.size] = self.items[self.capacity - 1];
        }
    }

    pub fn remove_by_value(&mut self, v: &T) {
        let idx = self.index_of_ref(v);
        if idx != -1 {
            self.remove_by_idx(idx as usize);
        }
    }

    pub fn is_empty(&self) -> bool {
        self.size == 0
    }

    pub fn contains(&self, v: T) -> bool {
        self.index_of_val(v) != -1
    }

    pub fn first(&self) -> Option<T> {
        if self.is_empty() {
            None
        } else {
            Some(self.items[0])
        }
    }

    pub fn last(&self) -> Option<T> {
        if self.is_empty() {
            None
        } else {
            Some(*self.items.last().unwrap())
        }
    }

    pub fn set(&mut self, idx: usize, v: &T) {
        if idx <= self.size {
            self.items[idx] = *v;
        }
    }

    pub fn capacity(&self) -> usize {
        self.capacity
    }
}
